File: let-nodejs/input/gps.js
#!/usr/bin/env node
/*  Goal-driven Parallel Sequences – Node.js port
    Original idea: Jos De Roo (Prolog)
    Python -> JS translation by ChatGPT
*/

/* ──────────────────────────────────────────────────────────
   Helpers (Prolog-style unification & instantiation)
   ────────────────────────────────────────────────────────── */
function isVar(token) {
  return typeof token === 'string' && (token[0] === '_' || /[A-Z]/.test(token[0]));
}

function unify(pattern, fact, env = {}) {
  if (pattern.length !== fact.length) return null;
  const e = { ...env };
  for (let i = 0; i < pattern.length; i++) {
    const p = pattern[i];
    const f = fact[i];
    if (isVar(p)) {
      if (p in e && e[p] !== f) return null;
      e[p] = f;
    } else if (p !== f) {
      return null;
    }
  }
  return e;
}

function instantiate(pattern, env) {
  return pattern.map(tok => (tok in env ? env[tok] : tok));
}

function factEquals(a, b) {
  if (a.length !== b.length) return false;
  return a.every((v, i) => v === b[i]);
}

/* ──────────────────────────────────────────────────────────
   Domain object
   ────────────────────────────────────────────────────────── */
class Transition {
  constructor(
    mapId,
    fromPattern,
    toPattern,
    action,
    duration,
    cost,
    belief,
    comfort
  ) {
    Object.assign(this, {
      mapId,
      fromPattern,
      toPattern,
      action,
      duration,
      cost,
      belief,
      comfort,
    });
  }
}

/* ──────────────────────────────────────────────────────────
   Utility predicates
   ────────────────────────────────────────────────────────── */
function stageCount(history) {
  if (history.length === 0) return 1;
  let stages = 1;
  for (let i = 1; i < history.length; i++) {
    if (history[i] !== history[i - 1]) stages += 1;
  }
  return stages;
}

function goalSatisfied(state, goalPattern) {
  return state.some(fact => unify(goalPattern, fact) !== null);
}

function applyTransition(state, trans, env) {
  const fromFact = instantiate(trans.fromPattern, env);
  const toFact = instantiate(trans.toPattern, env);

  const next = state.filter(f => !factEquals(f, fromFact)); // remove one
  next.push(toFact);
  return next;
}

function stateSignature(state) {
  // stable string key for loop-checking
  return state
    .map(f => JSON.stringify(f))
    .sort()
    .join('|');
}

/* ──────────────────────────────────────────────────────────
   Depth-first search generator
   ────────────────────────────────────────────────────────── */
function* dfs(
  state,
  transitions,
  goalPattern,
  limits,
  mapHist,
  pathHist,
  dur,
  cost,
  belief,
  comfort,
  visited
) {
  const sig = stateSignature(state);
  if (visited.has(sig)) return;
  visited.add(sig);

  if (goalSatisfied(state, goalPattern)) {
    yield [pathHist, dur, cost, belief, comfort];
  }

  const [maxDur, maxCost, minBelief, minComfort, maxStages] = limits;

  for (const tr of transitions) {
    if (stageCount([...mapHist, tr.mapId]) > maxStages) continue;

    for (const fact of state) {
      const env = unify(tr.fromPattern, fact);
      if (!env) continue;

      const dur2 = dur + tr.duration;
      if (dur2 > maxDur) continue;
      const cost2 = cost + tr.cost;
      if (cost2 > maxCost) continue;
      const belief2 = belief * tr.belief;
      if (belief2 < minBelief) continue;
      const comfort2 = comfort * tr.comfort;
      if (comfort2 < minComfort) continue;

      const nextState = applyTransition(state, tr, env);

      yield* dfs(
        nextState,
        transitions,
        goalPattern,
        limits,
        [...mapHist, tr.mapId],
        [...pathHist, tr.action],
        dur2,
        cost2,
        belief2,
        comfort2,
        new Set(visited) // local copy for back-tracking
      );
    }
  }
}

function* allPaths(state, transitions, goalPattern, limits) {
  yield* dfs(state, transitions, goalPattern, limits, [], [], 0, 0, 1, 1, new Set());
}

/* ──────────────────────────────────────────────────────────
   Example knowledge base (partial Belgium map)
   ────────────────────────────────────────────────────────── */
const transitions = [
  new Transition(
    'map_be',
    ['location', 'S', 'gent'],
    ['location', 'S', 'brugge'],
    'drive_gent_brugge',
    1500.0,
    0.006,
    0.96,
    0.99
  ),
  new Transition(
    'map_be',
    ['location', 'S', 'gent'],
    ['location', 'S', 'kortrijk'],
    'drive_gent_kortrijk',
    1600.0,
    0.007,
    0.96,
    0.99
  ),
  new Transition(
    'map_be',
    ['location', 'S', 'kortrijk'],
    ['location', 'S', 'brugge'],
    'drive_kortrijk_brugge',
    1600.0,
    0.007,
    0.96,
    0.99
  ),
  new Transition(
    'map_be',
    ['location', 'S', 'brugge'],
    ['location', 'S', 'oostende'],
    'drive_brugge_oostende',
    900.0,
    0.004,
    0.98,
    1.0
  ),
];

const initialState = [['location', 'i1', 'gent']];
const goalPattern = ['location', '_', 'oostende'];
const limits = [5000.0, 5.0, 0.2, 0.4, 1];

/* ──────────────────────────────────────────────────────────
   Demo driver (mirrors the original Prolog query)
   ────────────────────────────────────────────────────────── */
if (require.main === module) {
  let idx = 1;
  for (const [path, dur, cost, bel, comf] of allPaths(
    initialState,
    transitions,
    goalPattern,
    limits
  )) {
    console.log(`Solution ${idx++}`);
    console.log('  Path     :', path.join(' → '));
    console.log('  Duration :', dur);
    console.log('  Cost     :', cost);
    console.log('  Belief   :', bel);
    console.log('  Comfort  :', comf);
    console.log();
  }
}

File: let-python/input/gps.py
#!/usr/bin/env python3
"""
Goal-driven Parallel Sequences – full Python implementation
Based on the original Prolog program by Jos De Roo
Paper: https://www.sciencedirect.com/science/article/pii/S1532046421000794
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict, Generator, List, Optional, Set, Tuple


# ──────────────────────────────────────────────────────────────
#  Prolog-style helpers
# ──────────────────────────────────────────────────────────────
def _is_var(token: Any) -> bool:
    """Heuristic: a *variable* starts with a capital letter or '_'."""
    return isinstance(token, str) and (token[0].isupper() or token[0] == "_")


def unify(
    pattern: Tuple[Any, ...],
    fact: Tuple[Any, ...],
    env: Optional[Dict[str, Any]] = None
) -> Optional[Dict[str, Any]]:
    """
    Unify *pattern* with *fact* (1-to-1 correspondence).
    Return an extended binding environment or None on failure.
    """
    if len(pattern) != len(fact):
        return None

    env = dict(env or {})          # local copy
    for p, f in zip(pattern, fact):
        if _is_var(p):             # variable
            if p in env and env[p] != f:
                return None        # conflict
            env[p] = f
        elif p != f:               # constant mismatch
            return None
    return env


def instantiate(pattern: Tuple[Any, ...], env: Dict[str, Any]) -> Tuple[Any, ...]:
    """Replace variables in *pattern* using *env* bindings."""
    return tuple(env.get(tok, tok) for tok in pattern)


# ──────────────────────────────────────────────────────────────
#  Domain entities
# ──────────────────────────────────────────────────────────────
@dataclass(frozen=True)
class Transition:
    map_id: str
    from_pattern: Tuple[Any, ...]
    to_pattern: Tuple[Any, ...]
    action: str
    duration: float
    cost: float
    belief: float
    comfort: float


# ──────────────────────────────────────────────────────────────
#  Utility predicates
# ──────────────────────────────────────────────────────────────
def stage_count(maps_history: List[str]) -> int:
    """
    Count *stages* (= maximal contiguous sequences executed on the same map).
    Mirrors the Prolog stagecount/2 predicate.
    """
    if not maps_history:
        return 1
    stages = 1
    last = maps_history[0]
    for m in maps_history[1:]:
        if m != last:
            stages += 1
            last = m
    return stages


def goal_satisfied(state: Set[Tuple[Any, ...]], goal_pattern: Tuple[Any, ...]) -> bool:
    """True when *goal_pattern* unifies with at least one fact in *state*."""
    return any(unify(goal_pattern, fact) is not None for fact in state)


def apply_transition(
    state: Set[Tuple[Any, ...]],
    trans: Transition,
    env: Dict[str, Any]
) -> Set[Tuple[Any, ...]]:
    """
    Successor state after executing *trans* with binding *env*.
    Implements Prolog’s becomes/2 (replacement of one fact by another).
    """
    new_state = set(state)
    new_state.discard(instantiate(trans.from_pattern, env))
    new_state.add   (instantiate(trans.to_pattern,   env))
    return new_state


# ──────────────────────────────────────────────────────────────
#  Depth-first search (generator version)
# ──────────────────────────────────────────────────────────────
def _dfs(
    state: Set[Tuple[Any, ...]],
    transitions: List[Transition],
    goal_pattern: Tuple[Any, ...],
    limits: Tuple[float, float, float, float, int],
    maps_hist: List[str],
    path_hist: List[str],
    dur: float,
    cost: float,
    belief: float,
    comfort: float,
    visited: Set[frozenset],
) -> Generator[Tuple[List[str], float, float, float, float], None, None]:
    """
    Recursive DFS with pruning.  Yields every admissible (path, metrics) tuple.
    """
    state_key = frozenset(state)
    if state_key in visited:
        return
    visited.add(state_key)

    if goal_satisfied(state, goal_pattern):
        yield path_hist, dur, cost, belief, comfort

    max_dur, max_cost, min_belief, min_comfort, max_stages = limits

    for tr in transitions:
        # stage-limit
        if stage_count(maps_hist + [tr.map_id]) > max_stages:
            continue

        # try to fire the transition from any matching fact
        for fact in state:
            env = unify(tr.from_pattern, fact)
            if env is None:
                continue

            # metric updates & pruning
            dur2 = dur + tr.duration
            if dur2 > max_dur:
                continue
            cost2 = cost + tr.cost
            if cost2 > max_cost:
                continue
            belief2 = belief * tr.belief
            if belief2 < min_belief:
                continue
            comfort2 = comfort * tr.comfort
            if comfort2 < min_comfort:
                continue

            next_state = apply_transition(state, tr, env)

            yield from _dfs(
                next_state,
                transitions,
                goal_pattern,
                limits,
                maps_hist + [tr.map_id],
                path_hist + [tr.action],
                dur2,
                cost2,
                belief2,
                comfort2,
                visited.copy()        # local back-tracking
            )


def all_paths(
    state: Set[Tuple[Any, ...]],
    transitions: List[Transition],
    goal_pattern: Tuple[Any, ...],
    limits: Tuple[float, float, float, float, int],
) -> Generator[Tuple[List[str], float, float, float, float], None, None]:
    """
    Public entry: yield *every* admissible plan (path, duration, cost, belief, comfort).
    """
    yield from _dfs(
        state,
        transitions,
        goal_pattern,
        limits,
        maps_hist=[],
        path_hist=[],
        dur=0.0,
        cost=0.0,
        belief=1.0,
        comfort=1.0,
        visited=set(),
    )


# ──────────────────────────────────────────────────────────────
#  Example knowledge base (partial Belgium map)
# ──────────────────────────────────────────────────────────────
transitions: List[Transition] = [
    Transition(
        "map_be",
        ("location", "S", "gent"),
        ("location", "S", "brugge"),
        "drive_gent_brugge",
        1500.0,
        0.006,
        0.96,
        0.99,
    ),
    Transition(
        "map_be",
        ("location", "S", "gent"),
        ("location", "S", "kortrijk"),
        "drive_gent_kortrijk",
        1600.0,
        0.007,
        0.96,
        0.99,
    ),
    Transition(
        "map_be",
        ("location", "S", "kortrijk"),
        ("location", "S", "brugge"),
        "drive_kortrijk_brugge",
        1600.0,
        0.007,
        0.96,
        0.99,
    ),
    Transition(
        "map_be",
        ("location", "S", "brugge"),
        ("location", "S", "oostende"),
        "drive_brugge_oostende",
        900.0,
        0.004,
        0.98,
        1.00,
    ),
]

initial_state: Set[Tuple[Any, ...]] = {("location", "i1", "gent")}
goal_pattern: Tuple[Any, ...] = ("location", "_", "oostende")  # Prolog-style variable "_"
limits: Tuple[float, float, float, float, int] = (5000.0, 5.0, 0.2, 0.4, 1)


# ──────────────────────────────────────────────────────────────
#  Demo driver (mirrors the original Prolog query)
# ──────────────────────────────────────────────────────────────
if __name__ == "__main__":
    for i, (path, dur, cost, bel, comf) in enumerate(
        all_paths(initial_state, transitions, goal_pattern, limits), start=1
    ):
        print(f"Solution {i}")
        print("  Path     :", " → ".join(path))
        print("  Duration :", dur)
        print("  Cost     :", cost)
        print("  Belief   :", bel)
        print("  Comfort  :", comf)
        print()

